### PR TITLE
Adding dell ome device_info module

### PR DESCRIPTION
**Pull request by Sajna-Shetty**
##### SUMMARY
Submitting dellemc_ome_device_facts module for contribution

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
dellemc_ome_device_facts

##### ADDITIONAL INFORMATION
```
    "device_facts": [
            {
                "Actions": null,
                "AssetTag": null,
                "ChassisServiceTag": null,
                "ConnectionState": true,
                "DeviceManagement": [
                    {
                        "DnsName": "dnsname.host.com",
                        "InstrumentationName": "MX-12345",
                        "MacAddress": "11:10:11:10:11:10",
                        "ManagementId": 12345,
                        "ManagementProfile": [
                            {
                                "HasCreds": 0,
                                "ManagementId": 12345,
                                "ManagementProfileId": 12345,
                                "ManagementURL": "https://192.168.0.1:443",
                                "Status": 1000,
                                "StatusDateTime": "2019-01-21 06:30:08.501"
                            }
                        ],
                        "ManagementType": 2,
                        "NetworkAddress": "192.168.0.1"
                    }
                ],
                "DeviceName": "MX-0003I",
                "DeviceServiceTag": "MXL1234",
                "DeviceSubscription": null,
                "LastInventoryTime": "2019-01-21 06:30:08.501",
                "LastStatusTime": "2019-01-21 06:30:02.492",
                "ManagedState": 3000,
                "Model": "PowerEdge MX7000",
                "PowerState": 17,
                "SlotConfiguration": {},
                "Status": 4000,
                "SystemId": 2031,
                "Type": 2000
            }
        ]
```

### CODE REVIEW

File: lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py
Comment (rajeevarakkal) [inline lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py]: Is this empty line required?

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: normally it should be one space after colon

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: normally it should be one space after colon.

File: lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py
Comment (rajeevarakkal) [inline lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py]: isn't better to follow the format string style?

File: lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py
Comment (rajeevarakkal) [inline lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py]: Do you really using this main call?

Comment (Sajna-Shetty) [reply lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py]: Updated the code

File: lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py
Comment (Sajna-Shetty) [reply lib/ansible/module_utils/remote_management/dellemc/dellemc_ome.py]: Updated the code

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: Updated the code

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: Updated the code

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: Do you want to register an error details with `ansible_facts`?

Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: updated code by removing ansible_facts return in error case

**Review by rajeevarakkal — APPROVED** [review]
good.

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by rajeevarakkal** [comment]
shipit

**Review by rajeevarakkal — APPROVED** [review]

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by Sajna-Shetty** [comment]
ready_for_review

**Comment by Sajna-Shetty** [comment]
@gundalow 
It would be great if you could review this module

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by bcoca** [comment]
modules that return `ansible_facts` as their main function should be named with a `_facts` suffix

**Comment by felixfontein** [comment]
@bcoca the module has a `_facts` suffix.

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: I don't think it is clear from this description under which name the result is stored in `ansible_facts`. Maybe you should add an example which prints out the returned facts.

**Review by felixfontein — COMMENTED** [review]
You're currently storing the facts under hostname, which in all your examples is an IP address. This means it must be accessed via `ansible_facts['192.168.0.1']`. If the hostname happens to be a Python identifier, say `example`, users can access it directly as `example`.

Maybe it's better to make this an `_info` module which doesn't return `ansible_facts`, but where users have to `register:` the result. Then users can assign any name they want to the result.

File: lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py
Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: Thanks for the review. I have updated the description and example part.

Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: @felixfontein As we are connecting to remote system, we are retrieving unique information like network address,mac address and hardware details. As per [ansible doc](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variables-discovered-from-systems-facts), facts are used when connecting to Remote system,

So is not it better to have it as `_facts` module?

Comment (felixfontein) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: I don't know. I can think of reasons why it should be `_facts`, and reasons why it should be `_info`. I guess this is a question for the public Ansible meeting...

Comment (felixfontein) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: Discussion from meeting: should be `_facts`

Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/dellemc_ome_device_facts.py]: Thank you

**Comment by Sajna-Shetty** [comment]
As per meeting discussion, removed _dellemc__ from module name

**Comment by Sajna-Shetty** [comment]
@bcoca 
It would be great if you assign reviewer for this module

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by Sajna-Shetty** [comment]
@gundalow As you have reviewed earlier dellemc modules, Could you help us to get this module reviewed?

**Comment by Sajna-Shetty** [comment]
ready_for_review

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_facts.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_facts.py]: Sorry for bringing this up again. Modules should only be called `_facts` (and return `ansible_facts`) if these facts are about the machine the module runs on. This is not the case here, as the facts returned are about the host specified in `hostname`.

Can you change this line as follows:
```suggestion
            module.exit_json(device_info=device_facts)
```
and rename the module to `ome_device_info`?

(@bcoca promised to update the docs so that they are more clear on this topic.)

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_facts.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_facts.py]:
```suggestion
device_info:
```

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_facts.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_facts.py]:
```suggestion
  description: Returns the information collected from the Device.
```

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:167]:
```suggestion
  sample: |
    {
```
That way the YAML parser should treat the whole JSON block correctly as a string.

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (ghost) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py]:
```suggestion
            module.exit_json(device_info=device_facts)
```
Some of the sanity checks are failing because of this space. (ansibot should post the sanity test results, but apparently it isn't...)

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (ghost) [reply lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:167]: It still complains about invalid YAML at this position:
```
01:14 ERROR: Found 1 validate-modules issue(s) which need to be resolved:
01:14 ERROR: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:168:9: E313 RETURN is not valid YAML (100%)
```
https://app.shippable.com/github/ansible/ansible/runs/118514/1/console

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py]: @felixfontein thanks for the review. I will update the changes and update now

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:405]: In the sample for `RESULT`, you have `DeviceServiceTag` and not `device_service_tag`. Is that intentional?

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (felixfontein) [inline lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:168]: From how `device_facts` is treated in the module's main code, it does't look like `"value"` with a list is there, but that the contents of `device_info` are directly
```
        {
            "Actions": null,
            "AssetTag": null,
            "ChassisServiceTag": null,
            "ConnectionState": true,
            ...
```

Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:168]: The response returned from REST api contains key `value` with list of device information.

Comment (felixfontein) [reply lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:168]: Ok, just wanted to make sure that the docs are ok. I can't run the module, so I have to guess from the code if the docs are ok :)

**Review by felixfontein — APPROVED** [review]
shipit

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:405]: We have followed snake case for all options so `device_service_tag`  is used from module.`DeviceServiceTag ` is response returned from REST api

File: lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py
Comment (Sajna-Shetty) [reply lib/ansible/modules/remote_management/dellemc/ome/ome_device_info.py:168]: yes doc is proper : )

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by felixfontein** [comment]
@Sajna-Shetty you need more `shipit`'s for this to proceed. (`shipit`'s before the last push aren't counted.) It would also be helpful if someone could test this module independently (i.e. that person isn't simply executing a playbook you give that person, but tries to use the module without much help) and post their results here (like "I tested it with this playbook and it worked: [...]").

**Comment by rajeevarakkal** [comment]
shipit

**Comment by Sajna-Shetty** [comment]
> @Sajna-Shetty you need more `shipit`'s for this to proceed. (`shipit`'s before the last push aren't counted.) It would also be helpful if someone could test this module independently (i.e. that person isn't simply executing a playbook you give that person, but tries to use the module without much help) and post their results here (like "I tested it with this playbook and it worked: [...]").

Okay.Thanks

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by Sajna-Shetty** [comment]
bot_status

**Review by rajeevarakkal — APPROVED** [review]

**Review by rajeevarakkal — APPROVED** [review]
shipit

**Comment by rajeevarakkal** [comment]
@felixfontein, @Sajna-Shetty  I have tested this module and is working fine. Below is the basic play I have invoked which returned my all devices currently present in ome.

**Playbook**
```
---
- hosts: myome
  connection: local
  name: my dell system information from ome.
  gather_facts: no

  tasks:
  - name: retrieve my all devices.
    ome_device_info:
        hostname: "omeip"
        username: "omeuser"
        password: "omepass"
```

**Review by felixfontein — APPROVED** [review]
shipit

Potential problem before merging: the new module_utils module has GPLv3+ license, not BSD.

**Comment by Sajna-Shetty** [comment]
@felixfontein as this module meant only for Dell hardware's, we have used GPLv3 + license as per below guidline,
[https://docs.ansible.com/ansible/latest/dev_guide/developing_module_utilities.html]()

**Comment by bcoca** [comment]
https://docs.ansible.com/ansible/latest/dev_guide/developing_module_utilities.html#standard-module-utilities <=The first point shows GPL for modules, the second point shows module_util requires BSD

**Comment by Sajna-Shetty** [comment]
@felixfontein As per ansible guidelines we have updated utility file license to BSD

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by felixfontein** [comment]
@Sajna-Shetty I asked on #ansible-dev about BSD 3-clause, but I couldn't find anything out. It's probably a good idea to take this to the public Ansible meeting.

**Comment by Sajna-Shetty** [comment]
> @Sajna-Shetty I asked on #ansible-dev about BSD 3-clause, but I couldn't find anything out. It's probably a good idea to take this to the public Ansible meeting.

Thank you @felixfontein .We will attend meeting and get clarified.

**Comment by Sajna-Shetty** [comment]
@felixfontein Core team decided to use BSD 2-clause license for utility file. They will update the ansible doc for using specific BSD license.
Code has been updated to use BSD 2-Clause version.

**Comment by felixfontein** [comment]
@Sajna-Shetty I'm glad you could convince your legal team to accept the 2-clause BSD license!

I've checked that the license in the module_utils file equals the one in https://github.com/ansible/ansible/blob/devel/licenses/simplified_bsd.txt, so everything should be fine now! So it's time for:

shipit

**Comment by Sajna-Shetty** [comment]
> @Sajna-Shetty I'm glad you could convince your legal team to accept the 2-clause BSD license!
> 
> I've checked that the license in the module_utils file equals the one in https://github.com/ansible/ansible/blob/devel/licenses/simplified_bsd.txt, so everything should be fine now! So it's time for:
> 
> shipit

Thanks @felixfontein. So what is next step to merge this code?

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by Sajna-Shetty** [comment]
@felixfontein We have added unit test support now.Could you please check it and help us in merging this code?

**Comment by Sajna-Shetty** [comment]
bot_status

**Review by felixfontein — APPROVED** [review]
Thanks for adding the tests!

shipit

**Comment by felixfontein** [comment]
If nobody complains, I'll merge it on Monday.

**Comment by Sajna-Shetty** [comment]
> If nobody complains, I'll merge it on Monday.

Thank you :)

**Comment by Sajna-Shetty** [comment]
bot_status

**Comment by felixfontein** [comment]
rebuild_merge

**Comment by felixfontein** [comment]
It got merged, and will be available in Ansible 2.9! Congrats @Sajna-Shetty! I'm sorry it took so long...

**Comment by Sajna-Shetty** [comment]
> It got merged, and will be available in Ansible 2.9! Congrats @Sajna-Shetty! I'm sorry it took so long...

Thanks for taking your time to review 👍 and merging it :)

**Comment by rajeevarakkal** [comment]
Thanks @felixfontein, @Sajna-Shetty  and Ansible Core Team

**Comment by fbecker** [comment]
Is there a reason why this module is located inside a separate folder inside of remote_management/dellemc? Just wondering since the developer guide states that modules should not be nested more than two folders deep.

**Comment by fbecker** [comment]
Just noticed that the documentation page for the remote_management category is not displaying anything under dellemc. Maybe above is the reason for this?

**Comment by felixfontein** [comment]
You're right, I think it should be one level up. I'll create a PR to fix this.

**Comment by felixfontein** [comment]
(Same applies to the iDRAC modules in remote_management/dellemc/idrac.)

**Pull request merged**
